net: Add support for DSCP/ECN IPv4/IPv6 header fields

rlubos opened the pull request:
Currently there's no way to specify the value of DSCP/ECN IPv4/IPv6 header fields for outgoing packets, as well as those fields are not parsed from incoming packets. This PR attempts to fix this, by introducing the following changes:

* Specify the DSCP/ECN values at net_pkt level,
* Add processing of those fields in IPv4/IPv6 modules,
* Add new options for net_context and socket, which allow to specify the fields value on a net_context/socket,
* Add support for setting the fields value at ICMP level,
* Finally, introduce new options for `net shell` and `zperf` commands, which allow to configure DSCP/ECN value for outgoing packets.

Resolves #50139

rlubos committed:
net: pkt: Add IP DSCP/ECN information to a packet structure

Add new fields to the net_pkt structure, representing IP-specific
Differentiated services code point (DSCP) and Explicit Congestion
Notification (ECN) values. Those values are encoded in legacy
Type of Service (IPv4) and Trafic Class (IPv6) header fields.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv4: Add support for setting/reading DSCP/ECN fields (ToS)

Encode/decode recently added DSCP/ECN values into/from IPv4 header.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: ipv6: Add support for setting/reading DSCP/ECN fields (TC)

Encode/decode recently added DSCP/ECN fields into/from IPv6 header.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: context: Add DSCP/ECN support

Add a new net_context option, which allows to set DSCP/ECN values on a
net context. Those values are then encoded into outgoing packet
IPv4/IPv6 header.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: Add options to control DSCP/ECN value

Add new socket options IP_TOS and IPV6_TCLASS which allows to set
DSCP/ECN values on a socket for an outgoing packet IPv4/IPv6 headers.

The options are compatible with Linux behaviour, where both DSCP and ECN
are set with a single socket option.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: sockets: Add missing break statement in setsockopt

SOL_SOCKET and IPPROTO_TCP levels were missing the break statement at
the end of their processing logic, which could cause unexpected
fallthrough on unhandled optname value.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: icmpv4: Allow to specify DSCP and ECN values in ping

Add support for setting DSCP/ECN values for an outgoing ping request.
Additionally, copy DSCP/ECN values from an incoming ping request into
the ping response, like Linux does.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: icmpv6: Allow to specify DSCP and ECN values in ping

Add support for setting DSCP/ECN values for an outgoing ping request.
Additionally, copy DSCP/ECN values from an incoming ping request into
the ping response, like Linux does.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: shell: Add support for ping -Q parameter (Quality of Service)

Add new option to the net shell ping command, which allows to specify
the DSCP/ECN values for an outgoing ping request. The option is
compatible with Linux ping command.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: shell: Allow for ping arguments to be specified in hex

Improve the parse_arg() helper function used with shell ping command,
by allowing to parse arguments provided in hex.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: zperf: Add QoS support

Improve the zperf upload/upload2 commands, by allowing to specify
DSCP/ECP fields for outgoing packets. The introduced -S option is
compatible with Linux iperf3 utility.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@krish2718 Please have a look.

jukkar commented on subsys/net/ip/icmpv6.c:
```diff
@@ -358,6 +362,9 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 		return -ENOMEM;
 	}
 
+	net_pkt_set_ip_dscp(pkt, (tc & NET_IPV6_DSCP_MASK) >> NET_IPV6_DSCP_OFFSET);
+	net_pkt_set_ip_ecn(pkt, tc & NET_IPV6_ECN_MASK);
```
Same thing here

jukkar reviewed (commented):
Looks good, minor comment about API use.

rlubos commented on subsys/net/ip/icmpv4.c:
```diff
@@ -533,6 +537,9 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 		return -ENOMEM;
 	}
 
+	net_pkt_set_ip_dscp(pkt, (tos & NET_IPV4_DSCP_MASK) >> NET_IPV4_DSCP_OFFSET);
+	net_pkt_set_ip_ecn(pkt, tos & NET_IPV4_ECN_MASK);
```
Sure, that'd be cleaner, I'll provide helper functions.

krish2718 commented on subsys/net/ip/Kconfig:
```diff
@@ -628,6 +628,13 @@ config NET_CONTEXT_SNDBUF
 
+config NET_CONTEXT_DSCP_ECN
+	bool "Add support for setting DSCP/ECN IP properties on net_context"
+	help
```
IMHO we should make this default

rlubos replied:
Ok, no issues with that on my side.

krish2718 commented on subsys/net/ip/net_shell.c:
```diff
@@ -4084,7 +4088,7 @@ static int ping_ipv4(const struct shell *shell,
 #endif /* CONFIG_NET_IPV4 */
 
-static int parse_arg(size_t *i, size_t argc, char *argv[])
+static int parse_arg(size_t *i, size_t argc, char *argv[], bool allow_hex)
```
Do we really need this flag? We can allow HEX setting for options, no harm right?

rlubos replied:
Yeah, on second thought I think you're right.

rlubos commented:
@jukkar @krish2718 Thanks for the review. I've updated the PR with the following changes:
* Added IPv4/6 getter/setter for DSCP and ECN fields
* Made the net_context (and thus socket) options enabled by default
* dropped the unneeded `allow_hex` flag from `net ping` parser.

krish2718 reviewed (dismissed):
LGTM. Nit, IIUC, we are using `TOS` for applications and `DSCP_ECN` in networking stack, this is fine in general as ECN is only used by routers, but we could rename `TOS` to `DSCP_ECN` as well just to be consistent.

rlubos commented:
> Nit, IIUC, we are using TOS for applications and DSCP_ECN in networking stack, this is fine in general as ECN is only used by routers, but we could rename TOS to DSCP_ECN as well just to be consistent.

This is true, there is some variation with the naming, but here's my rationale:
* For the application side (ping/zperf/sockets) I've tried to keep compatibility with Linux nomenclature, to avoid confusion,
* For IPv4/IPv6 specific code, I sticked to the original field names, as the naming is already used in the header
* The generic part of network stack uses `dscp_ecn` as it should work with both IPv4/IPv6.

I hope that make sense.

tbursztyka reviewed (dismissed):
Please, make this optional (CONFIG_NET_IP_QOS ?). Not everyone is going to need this (by far), moreover it's adding 1 byte for each net_pkt.

rlubos commented:
@tbursztyka I've added `CONFIG_NET_IP_DSCP_ECN` to enable DSCP/ECN processing at IP and net_pkt level. I've left the option enabled by default.

rlubos commented:
@tbursztyka Can you please revisit?

rlubos commented:
@tbursztyka Please revisit.

jukkar reviewed (approved):
LGTM
